savaslabs/savaslabs.github.io
# Issue #2999: set up style guide

@oksana-c This PR does the following:

- Better organizes the gulpfile with comments that will be more apparent with code highlighting in the editor and a table of contents
- Adds the hologram gem, hologram config, and hologram assets (header and footer HTML, SCSS for the style guide, and logo/favicon images)
- Adds gulp tasks to build the style guide, compile style guide SCSS and pipe it to the style guide assets directory, and rebuild the style guide when changes are made
- Adds an introductory page for the style guide and a few sample components

The sample components will be updated when the existing SCSS files are removed or broken down into smaller components. They're just proof of concept at this point.

## To test

- Pull this branch
- Run `bundle install` to install the hologram gem
- Run `gulp serve`. Note: if you want this to go faster for the sake of testing the style guide, you can remove `'build:images', ` from line 360 of the gulpfile (the `build:local` task). There's an issue to only run unprocessed images through the image optimization task that I'd like to tackle this week if there's time.
- Once your local site pops up, you should be able to visit http://localhost:3000/styleguide/index.html
- Click on Components. Open up `_assets/styles/scss/components/_cards.scss` and try changing something in the SCSS comments above the `.card` styles, maybe the contents of `.card-header`. The style guide should refresh automatically to show you the change.
- Try adding a style like `body { background-color: blue; }` to one of the files in `_assets/styles/scss/styleguide`. The resulting CSS should be injected into the browser to show you the change.
- Try viewing the style guide at various screen sizes to ensure it looks good at all of them.
- Ctrl + C to quit `gulp serve`. Run `gulp clean`. Your `_site` directory should disappear. Run `gulp` to run the default `build` task (you may want to remove `'build:images', ` from line 334 of the gulpfile to make this quicker). Once this finishes, check out `_site/styleguide` to ensure the style guide directory has been built and contains an HTML file for each page, `styleguide.css`, and directories for images and js/styles from the main site.

## Still to do

- One thing I discovered late in the game when setting up hologram is that the resulting style guide is not themed at all 😹 Not ideal, but this does give us an opportunity to set up our own theming and branding. I'd like it to be fairly minimal in the end, but we could certainly do more to theme the styleguide to mimic the new website design, and add things like a mobile menu and more clearly delimited navigation on desktop. I already spent some personal hours this weekend theming the style guide for fun, but I don't have any more hours in this task to do more. We could consider adding a task to be completed down the road to improve the look and usability of the style guide. For now, it looks ok and is responsive.
- When we start refactoring the SCSS for the redesign, I'll be updating the style guide SCSS to pull in smaller components so there won't be as much duplication of code (e.g. I had to duplicate most of the header styles to override `_normalize.scss`)
- We'll need to add some styles to handle things like the color swatches you'll see when you test this thing out. This can be handled as part of component theming in my opinion.
- Travis should be able to install the hologram gem and build the style guide, but it's possible we will run into issues when we try to deploy this.

Please let me know if you see any errors or have any suggestions! :cat: :cat2: :octocat: :smile_cat:
